common-utils/src/main/java/eu/arrowhead/common/mqtt/handler/MqttTopicHandler.java

```diff
@@ -0,0 +1,233 @@
+		this.queue = queue;
+		this.threadpool = resourceManager.getThreadpool();
+		filters.sort((a, b) -> a.order() - b.order());
+	}
+
+	//-------------------------------------------------------------------------------------------------
+	@Override
+	public void run() {
+		Assert.isTrue(queue != null, getClass().getName() + "is not initialized");
+
+		doWork = true;
+		while (doWork) {
+			try {
+				final MqttMessage message = queue.take();
+				try {
+					threadpool.execute(() -> {
+						final long startTime = System.currentTimeMillis();
+						long endTime = 0;
+						MqttRequestModel request = null;
+						try {
+							final Entry<String, MqttRequestModel> parsed = parseMqttMessage(message);
+							request = parsed.getValue();
+
+							// Filter chain
+							for (final ArrowheadMqttFilter filter : filters) {
+								filter.doFilter(parsed.getKey(), request);
+							}
+
+							// API call
+							handle(request);
+
+						} catch (final Exception ex) {
+							errorResponse(ex, request);
+						} finally {
+							endTime = System.currentTimeMillis();
+							resourceManager.registerLatency(endTime - startTime);
+						}
+					});
+
+				} catch (final RejectedExecutionException ex) {
+					Entry<String, MqttRequestModel> parsed = null;
+					try {
+						parsed = parseMqttMessage(message);
+					} catch (final InvalidParameterException invalidEx) {
+						errorResponse(ex, null);
+						continue;
+					}
+					errorResponse(ex, parsed.getValue());
+				}
+
+			} catch (final InterruptedException ex) {
+				logger.debug(ex.getMessage());
+				logger.debug(ex);
+			}
+		}
+	}
+
+	//-------------------------------------------------------------------------------------------------
+	@Override
+	public void interrupt() {
+		doWork = false;
+		super.interrupt();
+	}
+
+	//-------------------------------------------------------------------------------------------------
+	public abstract String topic();
+
+	//-------------------------------------------------------------------------------------------------
+	public abstract void handle(final MqttRequestModel request) throws ArrowheadException;
+
+	//-------------------------------------------------------------------------------------------------
+	protected void successResponse(final MqttRequestModel request, final MqttStatus status, final Object response) {
+		if (!Utilities.isEmpty(request.getResponseTopic())) {
+			ahMqttService.response(Constants.MQTT_SERVICE_PROVIDING_BROKER_CONNECT_ID, request.getRequester(), request.getResponseTopic(), request.getTraceId(), request.getQosRequirement(), status, response);
+		} else {
+			logger.debug("No MQTT response topic was defined for success response.");
+		}
+	}
+
+	//-------------------------------------------------------------------------------------------------
+	protected <T> T readPayload(final Object payload, final Class<T> dtoClass) {
+		if (payload == null) {
+			return null;
+		}
+
```

Comment:
suggestion: maybe before we transform to string and back, we should try something like this:

```
if (dtoClass.isInstanceOf(payload)) {
    return (T) payload;
}
```